### PR TITLE
SYNTHESIZER: Add dump-loop-contracts into loop-contracts synthesizer

### PR DESCRIPTION
**Pull request by qinheping**
Running the loop-contracts synthesizer with the flag `dump-loop-contracts` will output the synthesized contracts as a JSON file, which can be used to annotate back to the source code with Crangler.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by jimgrundy — CHANGES_REQUESTED** [review]
I'm ok with the changes to the synthesizer, but the changes to c_wranger and the expression type not so much. They do look like improvements/fixes, but they should be factored into their own PRs so they and their consequences can be evaluated without holding up the synthesizer change.

Do please make those other PRs though.

File: src/goto-synthesizer/goto_synthesizer_parse_options.cpp [hunk 1]
```diff
@@ -47,23 +48,64 @@ int goto_synthesizer_parse_optionst::doit()
...
-    annotate_assigns(synthesizer.get_assigns_map(), goto_model);
-
-    // Apply loop contracts.
-    std::set<std::string> to_exclude_from_nondet_static(
-      cmdline.get_values("nondet-static-exclude").begin(),
-      cmdline.get_values("nondet-static-exclude").end());
-    code_contractst contracts(goto_model, log);
-    contracts.apply_loop_contracts(to_exclude_from_nondet_static);
+    // Default output destination is stdout.
+    std::string json_output_str = "stdout";
```
Comment (tautschnig) [inline src/goto-synthesizer/goto_synthesizer_parse_options.cpp:65]: This looked a bit weird to me on first sight, but am I right that this is understood by Crangler?

**Review by jimgrundy — APPROVED** [review]

File: src/goto-synthesizer/goto_synthesizer_parse_options.cpp (hunk 1 above)
Comment (qinheping) [reply src/goto-synthesizer/goto_synthesizer_parse_options.cpp:65]: Yes. https://github.com/diffblue/cbmc/blob/develop/src/crangler/c_wrangler.cpp#L616

Comment (tautschnig) [reply src/goto-synthesizer/goto_synthesizer_parse_options.cpp:65]: Would you mind just adding a comment to make sure others also understand this relationship to Crangler.

**Comment by tautschnig** [comment]
With all other PRs now merged: can you use expr2c with a custom configuration that sets up true and false as you initially did in the default configuration? I believe this will (after a rebase) make all tests pass.

**Review by tautschnig — APPROVED** [review]

**Pull request merged**